### DISCUSSION TITLE
Streaming Lumify Ultrasound Video to 3D Slicer via Plus Toolkit

### DISCUSSION THREAD

**Discussion by lagaron** [PlusToolkit/PlusLib#1193] [Q&A]
I am currently working on a project where I need to stream video from a Lumify ultrasound transducer to 3D Slicer in real-time. Here’s the setup and the issue I am facing:

1. Setup:
I am using a Samsung tablet running the Lumify app to capture ultrasound images. The tablet streams its screen using Screen Stream Mirroring to http://192.168.1.132:8080. And I want to use Plus Toolkit to acquire this video stream and send it to 3D Slicer.

2. Configuration Attempt:
I have created a configuration file for Plus Toolkit to acquire the video stream from the HTTP URL. I have tried using different device types (MediaFoundationVideo, MmfVideo, and OpenCVVideo) but the Plus Server stops unexpectedly with return code 1: Server stopped unexpectedly. Return code: 1

3. The XML I have tried with:
<PlusConfiguration version="2.4">
  <DataCollection StartupDelaySec="1.0">
    <DeviceSet Name="PlusServer: HTTP Video Stream" Description="Streaming video from HTTP source.">

      <Device Id="VideoDevice" Type="OpenCVVideo" Url="http://192.168.1.132:8080">
        <DataSources>
          <DataSource Type="Video" Id="VideoStream" PortUsImageOrientation="MF" ImageType="RGB_COLOR"/>
        </DataSources>
        <OutputChannels>
          <OutputChannel Id="VideoStreamChannel" VideoDataSourceId="VideoStream"/>
        </OutputChannels>
      </Device>

    </DeviceSet>
  </DataCollection>

  <PlusOpenIGTLinkServer MaxNumberOfIgtlMessagesToSend="1" MaxTimeSpentWithProcessingMs="50" ListeningPort="18944" SendValidTransformsOnly="true" OutputChannelId="VideoStreamChannel">
    <DefaultClientInfo>
      <MessageTypes>
        <Message Type="IMAGE"/>
      </MessageTypes>
      <ImageNames>
        <Image Name="VideoStream" EmbeddedTransformToFrame="Image"/>
      </ImageNames>
    </DefaultClientInfo>
  </PlusOpenIGTLinkServer>

</PlusConfiguration>
 
But I have tried others and I always get this error :( 

Any help or suggestions would be greatly appreciated!

Thank you in advance

**Comment by Sunderlandkyl** [comment]
Does the video stream from the tablet work with another application such as VLC?

**Reply by Sunderlandkyl** [reply]
Also the URL should be "VideoURL".

**Reply by lagaron** [reply]
I can't open it in VLC, but i can open it in my browser

**Reply by lassoan** [reply]
Id it works in your browser then it means the ultrasound provides a web server that hosts a web app that can receive and display a video stream. You need to replace that app with PLUS (or VLC, etc.). You can load the page with browser developer tools enabled to see what web API calls the web app performs to start streaming and what URL the video is received from. If some standard syreaming protocol is ueed then you may be able to replace the browser; otherwise you may need to pretend you are the browser to get access to the stream. You can also ask advice from Lumify, maybe they have an SDK that you can use.

**Reply by lagaron** [reply]
Now, I have achieved to see it on VLC too, with another URL. So now I am able to open it in both browser and VLC, however, in my XML code I still have the same error 1, when I try to run PlusServer. 
Here is the mentioned code:
![image](https://github.com/PlusToolkit/PlusLib/assets/78867544/b38332d6-8e80-40e9-a38f-c839cac7eb81)

**Reply by Sunderlandkyl** [reply]
Please attach the PlusServer log.
